### PR TITLE
Some refactoring, introduction of listify function.

### PR DESCRIPTION
**Pull request by jeandet**
Hello this PR isn't finished yet.
The idea is to discuss around some refactoring ideas.

First I would like to have some feedback on introducing a simple listify function to remove some 'noisy' code. If it's ok I'll convert more code.

Then maybe a more convenient function would directly parse kwargs, some extract function like this:

```python
def extract(dict_object, keys):
   return listify(*[dict_object.get(key,[]) for key in keys ])
```
So it could be called like this:

```
c_pchlist, cpp_pchlist = extract(kwargs,['c_pch', 'cpp_pch'])
```

### CODE REVIEW

File: mesonbuild/build.py [hunk 1]
```diff
@@ -607,64 +604,42 @@ def process_kwargs(self, kwargs, environment):
...
-        self.add_compiler_args('objcpp', objcpplist)
-        fortranlist = kwargs.get('fortran_args', [])
-        if not isinstance(fortranlist, list):
-            fortranlist = [fortranlist]
-        self.add_compiler_args('fortran', fortranlist)
-        rustlist = kwargs.get('rust_args', [])
-        if not isinstance(rustlist, list):
-            rustlist = [rustlist]
-        self.add_compiler_args('rust', rustlist)
+        compiler_args = {'c':clist, 'cpp':cpplist, 'cs':cslist, 'vala':valalist, 'objc':objclist, 'objcpp':objclist,
```
Comment (ignatenkobrain) [inline mesonbuild/build.py]: __[flake8]__
* [E231] missing whitespace after ':'

File: mesonbuild/build.py [hunk 2]
```diff
@@ -607,64 +604,42 @@ def process_kwargs(self, kwargs, environment):
...
+                         'fortran':fortranlist, 'rust':rustlist
```
Comment (ignatenkobrain) [inline mesonbuild/build.py]: __[flake8]__
* [E231] missing whitespace after ':'

File: mesonbuild/build.py [hunk 3]
```diff
@@ -607,64 +604,42 @@ def process_kwargs(self, kwargs, environment):
...
+                         }
+        for key,value in compiler_args.items():
```
Comment (ignatenkobrain) [inline mesonbuild/build.py]: __[flake8]__
* [E231] missing whitespace after ','

File: mesonbuild/build.py [hunk 4]
```diff
@@ -607,64 +604,42 @@ def process_kwargs(self, kwargs, environment):
...
+            self.add_compiler_args(key,value)
```
Comment (ignatenkobrain) [inline mesonbuild/build.py]: __[flake8]__
* [E231] missing whitespace after ','

File: mesonbuild/mesonlib.py [hunk 5]
```diff
@@ -470,6 +470,17 @@ def replace_if_different(dst, dst_tmp):
     else:
         os.unlink(dst_tmp)
 
+
+def listify(*args):
+    '''
+    Returns a list with all args embedded in a list if they are not of type list.
+    This function preserve order.
```
Comment (floppym) [inline mesonbuild/mesonlib.py]: Grammar mistake. Replace with:

This function preserves order.
or
This function will preserve order.

**Comment by jpakkane** [comment]
Looks good. Thanks for working on this. I should have fixed this ages ago but there's always "more important" stuff to do...

**Comment by nirbheek** [comment]
Indeed, thanks for looking into this :)

**Comment by jpakkane** [comment]
Is this ready to merge?

**Comment by jeandet** [comment]
I've rebased it this morning, there is still few places to refactor. I try to finish it for tomorrow morning.

File: mesonbuild/interpreter.py [hunk 6]
```diff
@@ -2374,9 +2356,7 @@ def add_test(self, node, args, kwargs, is_base_test):
         par = kwargs.get('is_parallel', True)
         if not isinstance(par, bool):
             raise InterpreterException('Keyword argument is_parallel must be a boolean.')
-        cmd_args = kwargs.get('args', [])
-        if not isinstance(cmd_args, list):
-            cmd_args = [cmd_args]
+        cmd_args = extract_as_list(kwargs ,'args')
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[flake8]__
* [E203] whitespace before ','
* [E231] missing whitespace after ','

**Comment by jpakkane** [comment]
Thanks a lot. This makes the code a lot nicer.

**Pull request merged**